### PR TITLE
add test/example for outbound request body streaming

### PR DESCRIPTION
**Pull request by dicej**

### CODE REVIEW

File: sdk/rust/src/http.rs [hunk 1]
```diff
@@ -461,14 +461,24 @@ impl IncomingResponse {
...
     ///
     /// Panics if the body was already consumed.
-    pub fn into_body_stream(self) -> impl futures::Stream<Item = Result<Vec<u8>, streams::Error>> {
+    // TODO: This should ideally take ownership of `self` and be called `into_body_stream` (i.e. symmetric with
+    // `IncomingRequest::into_body_stream`).  However, as of this writing, `wasmtime-wasi-http` is implemented in
+    // such a way that dropping an `IncomingResponse` will cause the request to be cancelled, meaning the caller
+    // won't necessarily have a chance to send the request body if they haven't started doing so yet (or, if they
+    // have started, they might not be able to finish before the connection is closed).  See
+    // https://github.com/bytecodealliance/wasmtime/issues/7413 for details.
+    pub fn take_body_stream(&self) -> impl futures::Stream<Item = Result<Vec<u8>, streams::Error>> {
```
Comment (itowlson) [inline sdk/rust/src/http.rs:470]: @dicej is this a planned change to the 2.0 SDK?

(Also, would this explain why your sample works and #1985 drops?)

File: examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs [hunk 2]
```diff
@@ -83,18 +89,121 @@ async fn handle_request(request: IncomingRequest, response_out: ResponseOutparam
...
-            response_out.set(response);
+            if let Some(url) = headers.iter().find_map(|(k, v)| {
+                (k == "url")
+                    .then_some(v)
+                    .and_then(|v| std::str::from_utf8(v).ok())
+                    .and_then(|v| Url::parse(v).ok())
+            }) {
+                match double_echo(request, &url).await {
+                    Ok((request_copy, response)) => {
+                        let mut stream = response.take_body_stream();
```
Comment (rylev) [inline examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs:103]: Finding it a bit hard to follow with the generic naming: perhaps using slightly more verbose names might help?

```suggestion
                        let mut incoming_response_body = response.take_body_stream();
```

File: examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs [hunk 3]
```diff
@@ -83,18 +89,121 @@ async fn handle_request(request: IncomingRequest, response_out: ResponseOutparam
...
+
+                        let response = OutgoingResponse::new(
+                            200,
+                            &Headers::new(
+                                &headers
+                                    .into_iter()
+                                    .filter_map(|(k, v)| (k == "content-type").then_some((k, v)))
```
Comment (rylev) [inline examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs:110]: Is this not equivalent?
```suggestion
                                    .filter(|(k, v)| k == "content-type")
```

File: examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs [hunk 4]
```diff
@@ -83,18 +89,121 @@ async fn handle_request(request: IncomingRequest, response_out: ResponseOutparam
...
+                                    .collect::<Vec<_>>(),
+                            ),
+                        );
+
+                        let mut body = response.take_body();
```
Comment (rylev) [inline examples/wasi-http-rust-streaming-outgoing-body/src/lib.rs:115]: Same as above

```suggestion
                        let mut outgoing_response_body = response.take_body();
```

File: sdk/rust/src/http.rs [hunk 5]
```diff
@@ -461,14 +461,24 @@ impl IncomingResponse {
     /// # Panics
     ///
     /// Panics if the body was already consumed.
-    pub fn into_body_stream(self) -> impl futures::Stream<Item = Result<Vec<u8>, streams::Error>> {
+    // TODO: This should ideally take ownership of `self` and be called `into_body_stream` (i.e. symmetric with
```
Comment (rylev) [inline sdk/rust/src/http.rs:464]: Sigh... I think this whole API might need another look - it's hard to get the flow right. Oh well...

File: sdk/rust/src/http.rs (hunk 1 above)
Comment (dicej) [reply sdk/rust/src/http.rs:470]: Yes and yes.  The API change is necessary because there's no other way to work around https://github.com/bytecodealliance/wasmtime/issues/7413, and I'm not sure when that will be fixed or whether it will be backported to Wasmtime 14 (and if it is, probably not in time for Spin 2.0).

Comment (dicej) [reply sdk/rust/src/http.rs:470]: Update: here's a PR with the fix: https://github.com/bytecodealliance/wasmtime/pull/7426
